### DISCUSSION TITLE
How do you track crashes/logs in an android SDK/library?

### DISCUSSION THREAD

**Discussion by youngatlas** [MobileNativeFoundation/discussions#174] [General]
Hi,
We are a small startup, and we are trying to build an app in the EV space. We have an android SDK with some limited functionality, which is currently being used in house. In the near future, we want to expose this SDK to our partners, and they will be using it too.
I wanted to add some kind of logging or crash reporting in our SDK to track the issues in it. The problem with all the logging/crash reporting libraries are that, you have to provide the configuration in the application class for android to initialize those libraries. We don't have an application class in our SDK, and I am not sure that every other client using our SDK will be happy to provide this configuration.
How do you track crashes/issues with the library/SDK which you distribute publically or to your clients?

**Comment by shahzadmajeed** [comment]
Probably just send everything to your backend and process there?

**Comment by philipphofmann** [comment]
The problem with this is that crash-reporting libraries usually need to hook into signal handlers, exception handlers, etc., which are global for the whole app. How would the crash reporting lib know to which library the crash belongs? So if you set up a crash reporting lib in your lib, you would get most likely crashes that are not even caused by you. 

What you could do is capture error logs and send all of them to your error reporting backend, but you need to make sure that your users are OK with this and be aware of personal data. You could achieve this with [Sentry](https://docs.sentry.io/platforms/android/), for example. Instead of installing it globally, you could interact with the Hub. You won't get hard crashes or any of the other features. Heads up, I work for Sentry, so I'm biased. Worth noting that we don't officially support that use case, see https://github.com/getsentry/sentry-docs/issues/3611, so you are on your own.

```Kotlin
val options = SentryOptions()
options.dsn = "your dsn"
val hub = Hub(options)
// sends your exception to your Sentry account
hub.captureException(throwable)
```

**Comment by ArielDemarco** [comment]
As was correctly pointed out, it is necessary to hook into the operating system's exception/signal handlers (whether on Android or iOS) to capture crashes. Unless your SDK is specifically for capturing crashes/exceptions, it is not recommended.

Regarding logging, as mentioned, you can use various pre-existing tools (e.g., Firebase / Sentry / Embrace / BugSnag / etc) that include this as part of their stack. However, something to keep in mind is that if the main application uses these SDKs, there could be potential dependency conflicts.

Personally, what I would advise is to have some in-house solution for reporting logs in critical or potentially problematic areas. Moreover, something that worked for us (as we also build an SDK) is to maintain constant communication with clients to determine if we're causing any kind of conflict.
